src/main/java/uk/gov/hmcts/reform/sscs/service/exceptions/FeignClientErrorDecoder.java

```diff
@@ -0,0 +1,89 @@
+package uk.gov.hmcts.reform.sscs.service.exceptions;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import feign.Request;
+import feign.Request.HttpMethod;
+import feign.Response;
+import feign.codec.ErrorDecoder;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.server.ResponseStatusException;
+import uk.gov.hmcts.reform.sscs.model.HmcFailureMessage;
+import uk.gov.hmcts.reform.sscs.model.single.hearing.HearingRequestPayload;
+import uk.gov.hmcts.reform.sscs.service.AppInsightsService;
+
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDateTime;
+
+@Slf4j
+public class FeignClientErrorDecoder implements ErrorDecoder {
+
+    private final AppInsightsService appInsightsService;
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    public FeignClientErrorDecoder(AppInsightsService appInsightsService) {
+        this.appInsightsService = appInsightsService;
+    }
+
+    @Override
+    public Exception decode(String methodKey, Response response) {
+        Request originalRequest = response.request();
+        HttpMethod httpMethod = originalRequest.httpMethod();
+        HmcFailureMessage failMsg = null;
+
+        if (httpMethod.equals(HttpMethod.POST) || httpMethod.equals(HttpMethod.PUT)) {
+            HearingRequestPayload payload = mapToPostHearingRequest(originalRequest);
+            if (payload != null) {
+                failMsg = HmcFailureMessage.builder()
+                    .requestType(httpMethod.toString())
+                    .caseID(Long.valueOf(payload.getCaseDetails().getCaseRef()))
+                    .timeStamp(LocalDateTime.now())
+                    .errorCode(String.valueOf(response.status()))
+                    .errorMessage(response.reason())
+                    .build();
+            }
+        } else {
+            Long caseId = Long.parseLong(response.request().requestTemplate().queries().get("id").iterator().next());
```

Comment:
How come you chose to use this method of getting the ID here? Is it safer?

Reply:
@dattipoeHMCTS Can you let me know another way to extract query params from Response/Request? This is the only way I know at the moment, thanks.